/* Copyright 2012 MaidSafe.net limited

This MaidSafe Software is licensed under the MaidSafe.net Commercial License, version 1.0 or later,
and The General Public License (GPL), version 3. By contributing code to this project You agree to
the terms laid out in the MaidSafe Contributor Agreement, version 1.0, found in the root directory
of this project at LICENSE, COPYING and CONTRIBUTOR respectively and also available at:

http://www.novinet.com/license

Unless required by applicable law or agreed to in writing, software distributed under the License is
distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
implied. See the License for the specific language governing permissions and limitations under the
License.
*/

package maidsafe.nfs.protobuf;

message DataName {
  required uint32 type = 1;
  required bytes raw_name = 2;
}

message ReturnCode {
  required int32 error_value = 1;
  required bytes error_category_name = 2;
}

message DataNameAndReturnCode {
  required bytes serialised_name = 1;
  required bytes serialised_return_code = 2;
}

message DataOrDataNameAndReturnCode {
optional bytes serialised_data_name_and_content = 1;
optional bytes serialised_name = 2;
optional bytes serialised_return_code = 3;
}

message DataNameAndVersion {
  required bytes serialised_data_name = 1;
  required bytes serialised_version_name = 2;
}

message DataNameVersionAndReturnCode {
  required bytes serialised_data_name_and_version = 1;
  required bytes serialised_return_code = 2;
}

message DataNameOldNewVersion {
  required bytes serialised_data_name = 1;
  required bytes serialised_old_version_name = 2;
  required bytes serialised_new_version_name = 3;
}

message DataNameOldNewVersionAndReturnCode {
  required bytes serialised_data_name_old_new_version = 1;
  required bytes serialised_return_code = 2;
}

message DataNameAndContent {
  required bytes serialised_name = 1;
  required bytes content = 2;
}

message DataAndReturnCode {
  required bytes serialised_data_name_and_content = 1;
  required bytes serialised_return_code = 2;
}

message DataAndPmidHint {
  required bytes serialised_data_name_and_content = 1;
  required bytes pmid_hint = 2;
}

message DataPmidHintAndReturnCode {
  required bytes serialised_data_and_pmid_hint = 1;
  required bytes serialised_return_code = 2;
}

message PmidRegistration {
  message SignedDetails {
    message Details {
      required bytes maid_name = 1;
      required bytes pmid_name = 2;
      required bool unregister = 3;
    }
    required bytes serialised_details = 1;
    required bytes pmid_signature = 2;
  }
  required bytes serialised_signed_details = 1;
  required bytes maid_signature = 2;
}

message PmidRegistrationAndReturnCode {
  required bytes serialised_pmid_registration = 1;
  required bytes serialised_return_code = 2;
}
